### PR TITLE
feat(@toss/react): Add createStyledComponent function

### PR DESCRIPTION
**Pull request by soonki-98**
## Overview
I worked on to create function named "`createStyledComponent`". This function is not only for create new styled component but also for support "`tag`" by tag props or `dot(.)`.

createStyledComponent support below functions.
1. create styled component with tag.
2. support Ref
3. can overrided by styled function that `@emotion/styled` package provide.
4. can use custom props

I suggest to use this function for common UI code like `flex`.

## Demo Code
```tsx
######## Define Component ########
const FlexStyledComponent = styled(
  forwardRef((props: { tag?: keyof JSX.IntrinsicElements } & FlexOptions & AllHTMLAttributes<HTMLElement>, ref) => {
    const { tag = 'div', ...rest } = props;

    return React.createElement(tag, { ...rest, ref });
  })
)`
  ${props => flex(props.align, props.justify, props.direction)}
`;

const FlexBase = forwardRef((props: FlexOptions & AllHTMLAttributes<HTMLElement>, ref) =>
  createStyledComponent(FlexStyledComponent)({ ...props, ref })
);

const Flex = FlexBase as CreateFlex;

jsxIntrinsicElements.forEach(element => {
  Flex[element] = forwardRef((props: FlexOptions & AllHTMLAttributes<HTMLElement>, ref) =>
    createStyledComponent(FlexStyledComponent, element)[element]({ ...props, ref })
  );
});

export default Flex;

######## Use Component ########
import { Flex } from './';
import styled from '@emotion/styled';

const FlexULList = styled(Flex.ul)``
const FlexLIItem = styled(Flex)``

const Page = () => {
   return (
     <FlexULList align="center">
          <FlexLIItem.li>hi</FlexLIItem.li>
          <Flex.li>hi</Flex.li>
     </FlexULList>
   )
}

```

## Comment
I wonder can I install new npm package in this package scope. Please review about this.

## PR Checklist

- [x] I read and included theses actions below

1. I have read the [Contributing Guide](https://github.com/toss/slash/blob/main/.github/CONTRIBUTING.md)
6. I have written documents and tests, if needed.

### CODE REVIEW

**Comment by ChanhyukPark-Tech** [comment]
Thank you for the great work.
I don't think I've got the full context of this PR.
**I wonder what are the advantages of this function `createStyledComponent` compared to the current structure**

**Comment by soonki-98** [comment]
Thanks for the feedback.

What I expect is `more extensibility` and `more convenience` when using `Flex component` for example.
Right now, if you want to use a Flex component with other tags, you have to use the prop `as`, so if you want to override a Flex component with styled features, you can't define the tags when overriding.
But because of `createStyledComponent` offers tags with `dot`, you can do just that. 

### Flex component right now
```typescript
const ULFlex = styled(Flex)``
<ULFlex.Center as="ul" />
```

### Flex component defined with `createStyledComponent`
```typescript
const ULFlex = styled(Flex.ul)``
<ULFlex.Center />
```

**Comment by HoseungJang** [comment]
I think this is unnecessarily complicated. So what are the advantages of after?

```tsx
// before
const StyledFlex = styled(Flex);
<StyledFlex as="ul"></StyledFlex>

// after
const FlexStyledComponent = styled(
  forwardRef((props: { tag?: keyof JSX.IntrinsicElements } & FlexOptions & AllHTMLAttributes<HTMLElement>, ref) => {
    const { tag = 'div', ...rest } = props;

    return React.createElement(tag, { ...rest, ref });
  })
)`
  ${props => flex(props.align, props.justify, props.direction)}
`;

const FlexBase = forwardRef((props: FlexOptions & AllHTMLAttributes<HTMLElement>, ref) =>
  createStyledComponent(FlexStyledComponent)({ ...props, ref })
);

const Flex = FlexBase as CreateFlex;

jsxIntrinsicElements.forEach(element => {
  Flex[element] = forwardRef((props: FlexOptions & AllHTMLAttributes<HTMLElement>, ref) =>
    createStyledComponent(FlexStyledComponent, element)[element]({ ...props, ref })
  );
});

const StyledFlex = styled(Flex.ul)``;

<StyledFlex></StyledFlex>
```

File: packages/react/emotion-utils/src/createStyledComponent.tsx [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+};
+export function createStyledComponent<P>(StyledBase: StyledBaseType<P>, tag?: keyof JSX.IntrinsicElements) {
+  if (!tag) {
+    return ({ ...rest }: React.ComponentProps<typeof StyledBase>) => <StyledBase {...rest} />;
+  } else {
+    return {
+      [tag]: ({ ...rest }: React.ComponentProps<typeof StyledBase>) => <StyledBase tag={tag} {...rest} />,
+    };
+  }
+}
```
Comment (HoseungJang) [inline packages/react/emotion-utils/src/createStyledComponent.tsx:23]: `StyledBaseType` doesn't guarantee that the returned element from the component is actually created as `tag`, for exmaple:

```tsx
const Component = styled((props: StyledBaseType<{}>) => {
  return createElement('div', props);
})``;

const ULComponent = createStyledComponent(Component, 'ul'); // but ULComponent is actually div
```

**Comment by HoseungJang** [comment]
First of all I think this is unnecessarily complicated. Basically what are the advantages of `after`?

```tsx
// before
const StyledFlex = styled(Flex);
<StyledFlex as="ul"></StyledFlex>

// after
const FlexStyledComponent = styled(
  forwardRef((props: { tag?: keyof JSX.IntrinsicElements } & FlexOptions & AllHTMLAttributes<HTMLElement>, ref) => {
    const { tag = 'div', ...rest } = props;

    return React.createElement(tag, { ...rest, ref });
  })
)`
  ${props => flex(props.align, props.justify, props.direction)}
`;

const FlexBase = forwardRef((props: FlexOptions & AllHTMLAttributes<HTMLElement>, ref) =>
  createStyledComponent(FlexStyledComponent)({ ...props, ref })
);

const Flex = FlexBase as CreateFlex;

jsxIntrinsicElements.forEach(element => {
  Flex[element] = forwardRef((props: FlexOptions & AllHTMLAttributes<HTMLElement>, ref) =>
    createStyledComponent(FlexStyledComponent, element)[element]({ ...props, ref })
  );
});

const StyledFlex = styled(Flex.ul)``;
<StyledFlex></StyledFlex>
```

**Pull request closed without merging**